### PR TITLE
Gigaspace metrics MBean

### PR DESCRIPTION
**Pull request by ath0s**
TP-897: When using Gigaspace modules, export an MBean with LRMI thread pool metrics.

### CODE REVIEW

File: astrix-context/src/main/java/com/avanza/astrix/context/mbeans/PlatformMBeanServer.java [hunk 1]
```diff
@@ -52,32 +51,40 @@ public void registerMBean(Object mbean, String folder, String name) {
...
 	}
-	
+
 	private void unregisterMBean(ObjectName objectName) {
 		try {
-			logger.debug("Unregister mbean: name={}", objectName);
-			ManagementFactory.getPlatformMBeanServer().unregisterMBean(objectName);
+			MBeanServer platformMBeanServer = ManagementFactory.getPlatformMBeanServer();
+			if (platformMBeanServer.isRegistered(objectName)) {
+				logger.debug("Unregister mbean: name={}", objectName);
```
Comment (askoog) [inline astrix-context/src/main/java/com/avanza/astrix/context/mbeans/PlatformMBeanServer.java:82]: 👍

File: astrix-context/src/main/java/com/avanza/astrix/context/mbeans/PlatformMBeanServer.java [hunk 2]
```diff
@@ -52,32 +51,40 @@ public void registerMBean(Object mbean, String folder, String name) {
...
+		exportedMbeans.forEach(this::unregisterMBean);
 	}
-	
+
 	private void unregisterMBean(ObjectName objectName) {
 		try {
-			logger.debug("Unregister mbean: name={}", objectName);
-			ManagementFactory.getPlatformMBeanServer().unregisterMBean(objectName);
+			MBeanServer platformMBeanServer = ManagementFactory.getPlatformMBeanServer();
+			if (platformMBeanServer.isRegistered(objectName)) {
```
Comment (askoog) [inline astrix-context/src/main/java/com/avanza/astrix/context/mbeans/PlatformMBeanServer.java:81]: 👍 nice!

File: astrix-gs/src/main/java/com/avanza/astrix/gs/GigaspaceMetricsExporterImpl.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+import com.avanza.astrix.beans.core.AstrixConfigAware;
+import com.avanza.astrix.beans.core.AstrixSettings;
+import com.avanza.astrix.config.DynamicBooleanProperty;
+import com.avanza.astrix.config.DynamicConfig;
+import com.avanza.astrix.context.mbeans.MBeanExporter;
+import com.avanza.astrix.gs.metrics.GigaspaceMetrics;
+
+final class GigaspaceMetricsExporterImpl implements AstrixConfigAware, GigaspaceMetricsExporter {
+	private static final String MBEAN_FOLDER = "Gigaspaces";
```
Comment (askoog) [inline astrix-gs/src/main/java/com/avanza/astrix/gs/GigaspaceMetricsExporterImpl.java:30]: I guess this means that the mbean path will be `com.avanza.astrix.context:00=Gigaspaces,name=Metrics`?

**Review by askoog — APPROVED** [review]

File: astrix-gs/src/main/java/com/avanza/astrix/gs/GigaspaceMetricsExporterImpl.java (hunk 3 above)
Comment (ath0s) [reply astrix-gs/src/main/java/com/avanza/astrix/gs/GigaspaceMetricsExporterImpl.java:30]: Yeah, exactly

**Review by Bjornelmers — APPROVED** [review]
Well done @ath0s!

**Pull request merged**